### ISSUE TITLE
Is it possible to create ipywidgets in a *.py file and return them to notebook to display?

### ISSUE THREAD

**Issue by sbonaretti** [jupyter-widgets/ipywidgets#2622]
Hi!

In one of my projects I use Jupyter notebooks as a user interface. The user runs cells that call (very long!) functions written in a python package made of `.py` files and sees output images in the notebook. I would love the user to be able to browse some of the output images using IntSlider. 

Is there a way I can write the function creating the sliders in a `.py` file and then return the sliders to notebook to display them? I have tried to write a function that returns HBox, but when called from the notebook, HBox = None (it is an actual object in the function in my `.py` file)

Thanks!!!

**Comment by jasongrout** [comment]
> Is there a way I can write the function creating the sliders in a `.py` file and then return the sliders to notebook to display them?

Yes.

>  I have tried to write a function that returns HBox, but when called from the notebook, HBox = None (it is an actual object in the function in my `.py` file)

Can you post a simplified version of what you have written, so we can understand what you know and have tried and to better understand exactly what you are trying to do?

**Comment by sbonaretti** [comment]
Thanks so much for replying so quickly!

I have found my bug! I was forgetting to return the box from an inner function to an outer one! I have to return it twice, because I have two nested functions. 
But now I have another issue with the same code, if I may:

I have medical images and I want to browse them using a slider. Because the `display` function for the image is in my `.py` file, I do not see the image but only the slider. Please, see the function below `browse_images`.

# In the `.py` file: 

# this is the function called by the notebook
def show_images():
      # Create a matplotlib figure
     ```
      fig1, ax1 = plt.subplots(1, 1);
      ```
      Pass fig and axis to the function that uses widgets:
      ```
      box = browse_images(img, fig1, ax1)
      return box
      ```

# this is the function that creates the widgets and it is called by the previous one:
```
def browse_images(img, fig1, ax1)

    # function for slider_1 
    def view_image_1(slider_1):
        ax1.imshow(img_orig_py[:,:,slider_1], cmap=plt.cm.gray, origin='lower',interpolation=None)
        ax1.set_title("Image title")
        ax1.axis('off')
        display(fig1) # <----- This function throws an error when in .py file, but works in notebook
        
    # link sliders and functions
    widget_1 = interactive(view_image_1, slider_1 = widgets.IntSlider(min=0, max=size[2]-1, value=slice_ID, description='Slice n.'))    
    
    # show figures before start interacting
    widget_1.update() 
    
    # create the vertical box for image 1
    controls_1 = HBox(widget_1.children[:-1]) # get the slider
    output_1   = widget_1.children[-1]              # get the image
    vBox_1 = VBox([controls_1, output_1])       # slider and image in a vertical box

    # return the vertical box 
    return vBox_1
```

In notebook, I have:
```
box = browse_images(img, fig1, ax1)
display(box)
```
If I comment out `display(fig1) ` I get the slider, but not the image. 

Can you recommend maybe a better way to do all this? In the picture below is what I can achieve directly in the notebook.

<img width="610" alt="working_on_notebook" src="https://user-images.githubusercontent.com/13563701/69350836-89765b00-0c7a-11ea-9a16-29c78c71fadc.png">

**Comment by jasongrout** [comment]
>            display(fig1) # <----- This function throws an error when in .py file, but works in notebook

What is the error?

**Comment by jasongrout** [comment]
Two thoughts:

1. If you use `plt.show()` (i.e., the matplotlib show function), does it work?

2. Did you know that matplotlib has a widgets frontend? In other words, you can make an mpl plot that actually is a widget and can be put in an hbox/vbox and controlled directly. See the https://github.com/matplotlib/jupyter-matplotlib package, and in particular see the last example in https://github.com/matplotlib/jupyter-matplotlib/blob/master/examples/ipympl.ipynb for how to display a slider and plot, where the slider controls the plot.

**Comment by sbonaretti** [comment]
It works!!! I am new here, and thanks to your inputs I could try a few things and make it work! Mainly knowing that it was possible, pushed me to actually find a solution! Thanks so much!!!

In my `.py` I imported `from ipywidget import display`, but I have just found that `display` is not part of `ipywidget`. I changed the import to `from ipywidget import *` and I everything is working!!! 

Hope I have not wasted too much of your time. Thanks!
